packages/inference/src/providers/fal-ai.ts

```diff
@@ -29,12 +37,85 @@ const makeHeaders = (params: HeaderParams): Record<string, string> => {
 };
 
 const makeUrl = (params: UrlParams): string => {
-	return `${params.baseUrl}/${params.model}`;
+	const baseUrl = `${params.baseUrl}/${params.model}`;
+	if (params.authMethod !== "provider-key" && params.task === "text-to-video") {
+		return `${baseUrl}?_subdomain=queue`;
+	}
+	return baseUrl;
 };
 
 export const FAL_AI_CONFIG: ProviderConfig = {
-	baseUrl: FAL_AI_API_BASE_URL,
+	makeBaseUrl,
 	makeBody,
 	makeHeaders,
 	makeUrl,
 };
+
+export interface FalAiQueueOutput {
+	request_id: string;
+	status: string;
+	response_url: string;
+}
+
+export async function pollFalResponse(
+	res: FalAiQueueOutput,
+	url: string,
+	headers: Record<string, string>
+): Promise<Blob> {
+	const requestId = res.request_id;
+	if (!requestId) {
+		throw new InferenceOutputError("No request ID found in the response");
+	}
+	let status = res.status;
+
+	const parsedUrl = new URL(url);
+	const baseUrl = `${parsedUrl.protocol}//${parsedUrl.host}${
+		parsedUrl.host === "router.huggingface.co" ? "/fal-ai" : ""
+	}`;
+
+	// extracting the provider model id for status and result urls
+	// from the response as it might be different from the mapped model in `url`
+	const modelId = new URL(res.response_url).pathname;
+	const queryParams = parsedUrl.search;
+
+	const statusUrl = `${baseUrl}${modelId}/status${queryParams}`;
+	const resultUrl = `${baseUrl}${modelId}${queryParams}`;
+
+	while (status !== "COMPLETED") {
+		await delay(500);
+		const statusResponse = await fetch(statusUrl, { headers: headers });
+
+		if (!statusResponse.ok) {
+			throw new InferenceOutputError("Failed to fetch response status from fal-ai API");
+		}
+		try {
+			status = (await statusResponse.json()).status;
+		} catch (error) {
+			throw new InferenceOutputError("Failed to parse status response from fal-ai API");
+		}
+	}
+
+	const resultResponse = await fetch(resultUrl, { headers: headers });
+	let result;
+	try {
+		result = await resultResponse.json();
+	} catch (error) {
+		throw new InferenceOutputError("Failed to parse result response from fal-ai API");
+	}
+	const isValidOutput =
+		typeof result === "object" &&
+		!!result &&
+		"video" in result &&
+		typeof result.video === "object" &&
+		!!result.video &&
+		"url" in result.video &&
+		typeof result.video.url === "string" &&
+		isUrl(result.video.url);
+	if (!isValidOutput) {
+		throw new InferenceOutputError(
+			"Expected { video: { url: string } } result format, got instead: " + JSON.stringify(result)
+		);
+	}
```

Comment:
Note: Moving the condition inside the `if` test will trigger Typescript type inference
```
	if (typeof result === "object" &&
		!!result &&
		"video" in result &&
		typeof result.video === "object" &&
		!!result.video &&
		"url" in result.video &&
		typeof result.video.url === "string" &&
		isUrl(result.video.url) {
			/// result has the correct type here
	} else {
		throw new InferenceOutputError(
			"Expected { video: { url: string } } result format, got instead: " + JSON.stringify(result)
		);
	}
```